Feature/401: add docs for mojaloop integration (#14)
mojaloop/pisp-demo-server: docs/design/mojaloop_integration.md

![Figure 1](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/stevenwjy/pisp-demo-server/mojaloop-integration-docs/docs/assets/diagrams/mojaloop/fig-01-traditional-method.puml)

![Figure 2](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/stevenwjy/pisp-demo-server/mojaloop-integration-docs/docs/assets/diagrams/mojaloop/fig-02-mojaloop-rtp.puml)

As RTP system like Mojaloop grows in popularity, the use of third-party services who do not manage 
its own ledger but exist for the purpose of initiating transactions has also come into attention.
With mojaloop, a PISP also enjoys the benefit of the FSPs where it does not need to mantain multiple
APIs to communicate with each FSP but only a single API that follows Mojaloop's standard. Figure 3
shows how a PISP could simultaneously setup communication with multiple FSPs by integrating to the
Mojaloop's RTP network.

![Figure 3](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/stevenwjy/pisp-demo-server/mojaloop-integration-docs/docs/assets/diagrams/mojaloop/fig-03-pisp-mojaloop.puml)

> In order to communicate with Mojaloop, below are the necessary setup required for a PISP:

Reviewer: ```suggestion
In order to join a Mojaloop network, a PISP must complete the following steps:
```

1. [Certificate Registration](#certificate-registration)

> 2. [URLs Registration](#url-registration)

Reviewer: ```suggestion
2. [URL Registration](#url-registration)
```

issued certificate, Mojaloop's URL, PISP identifier, transport scheme, and various other options
which could be found [here](https://github.com/mojaloop/sdk-standard-components/blob/master/src/lib/requests/baseRequests.js).

## Certificate Registration

In the Mojaloop's network for FSP Interoperability, every participant must be registered and have a
signed certificate by the centralized Certificate Authority (CA). According to the 
[PKI Best Practices](https://docs.mojaloop.io/mojaloop-specification/documents/PKI%20Best%20Practices.html),
the CA itself will have a self-signed root certificate for signing the participants' certificates. 
The certificates are important to protect the transport-level communication (TLS) and application-level
communication (JWS and JWE). All of the security protocols involved are for the purpose of ensuring
integrity and confidentiality between platforms. More information could be found in the 
[Mojaloop Specifications](https://docs.mojaloop.io/mojaloop-specification/).

PISP has to register its certificate manually, by telling the operator of Mojaloop the PISP's public
key and have it digitally signed by the centrallized CA.

> ## URLs Registration

Reviewer: ```suggestion
## URL Registration
```

As Mojaloop's architecture follows the asynchronous request/response design pattern, PISP demo server 
also interacts with Mojaloop using that pattern to avoid long-running connections. When making a 
request to Mojaloop, it may take some time to receive the response as everything is considered to work 
asynchronously. Even though it takes only several seconds on average, we still need to prepare for the 
indefinite time characteristic of asynchronous messages. As a result, PISP demo server needs to register 
its URL so that Mojaloop knows where to make the callback once the reply is ready. For example, 
Mojaloop needs to know the endpoint of the PISP server that is ready to handle `PUT /parties/{Type}/{ID}` 
callback from Mojaloop following a previous request of `GET /parties/{Type}/{ID}` to perform a party lookup.

The following are the callback URLs that need to be registered by the PISP demo server:

> - `PUT /participants`

Reviewer: This looks good to me - we are however missing the `PUT .../error` callbacks that must be registered for the switch to deliver error messages to the PISP, e.g. `PUT /parties/{Type}/{ID}/error` (see [here](https://docs.mojaloop.io/mojaloop-specification/documents/API%20Definition%20v1.0.html#6341-put-partiestypeiderror))

Author: Okay, will update soon

- `PUT /parties/{Type}/{ID}`
- `PUT /consentRequests/{ID}`

> - `POST /consents`

Reviewer 2: I might just be confused about functionality here but wouldn't we also need to mention the specific routes beyond the `.../{ID}/..` such as `POST consent/{ID}/revoke` or `POST consent/{ID}/generateChallenge` ?

Author: I think that's unnecessary here because those are the requests that we need to make to Mojaloop; meanwhile, the endpoints that we write down here are the callbacks that we expect Mojaloop to make to the PISP server. 

However, just realized that we haven't included the callback for consent revocation. Will update it
